diff -w -r --unified=5 apt-1.0.9.8.3.original/Makefile apt-1.0.9.8.3/Makefile
--- apt-1.0.9.8.3.original/Makefile	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/Makefile	2016-06-04 17:53:18.000000000 +0100
@@ -9,21 +9,22 @@
 .PHONY: default
 default: startup all

 .PHONY: headers library clean veryclean all binary program doc test update-po
 all headers library clean veryclean binary program doc manpages docbook test update-po startup dirs:
+	$(MAKE) -C darwin $@
 	$(MAKE) -C vendor $@
 	$(MAKE) -C apt-pkg $@
 	$(MAKE) -C apt-inst $@
 	$(MAKE) -C apt-private $@
 	$(MAKE) -C methods $@
 	$(MAKE) -C cmdline $@
 	$(MAKE) -C ftparchive $@
 	$(MAKE) -C dselect $@
-	$(MAKE) -C doc $@
+	#$(MAKE) -C doc $@
 	$(MAKE) -C po $@
-	$(MAKE) -C test $@
+	#$(MAKE) -C test $@

 all headers library clean veryclean binary program doc manpages docbook test update-po: startup dirs

 dirs: startup

diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/aptconfiguration.cc apt-1.0.9.8.3/apt-pkg/aptconfiguration.cc
--- apt-1.0.9.8.3.original/apt-pkg/aptconfiguration.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/aptconfiguration.cc	2016-06-04 12:29:03.000000000 +0100
@@ -191,11 +191,11 @@
 	}

 	// get the environment language codes: LC_MESSAGES (and later LANGUAGE)
 	// we extract both, a long and a short code and then we will
 	// check if we actually need both (rare) or if the short is enough
-	string const envMsg = string(Locale == 0 ? std::setlocale(LC_MESSAGES, NULL) : *Locale);
+	string const envMsg = string(Locale == 0 ? setlocale(LC_MESSAGES, NULL) : *Locale);
 	size_t const lenShort = (envMsg.find('_') != string::npos) ? envMsg.find('_') : 2;
 	size_t const lenLong = (envMsg.find_first_of(".@") != string::npos) ? envMsg.find_first_of(".@") : (lenShort + 3);

 	string const envLong = envMsg.substr(0,lenLong);
 	string const envShort = envLong.substr(0,lenShort);
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/contrib/cmndline.cc apt-1.0.9.8.3/apt-pkg/contrib/cmndline.cc
--- apt-1.0.9.8.3.original/apt-pkg/contrib/cmndline.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/contrib/cmndline.cc	2016-06-04 14:48:30.000000000 +0100
@@ -21,10 +21,14 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
 #include <string>

+#ifdef __APPLE__
+  #include "darwin/strchrnul.h"
+#endif // __APPLE__
+
 #include <apti18n.h>
 									/*}}}*/
 using namespace std;

 // CommandLine::CommandLine - Constructor				/*{{{*/
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/contrib/fileutl.cc apt-1.0.9.8.3/apt-pkg/contrib/fileutl.cc
--- apt-1.0.9.8.3.original/apt-pkg/contrib/fileutl.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/contrib/fileutl.cc	2016-06-04 13:22:56.000000000 +0100
@@ -58,11 +58,18 @@
 	#include <bzlib.h>
 #endif
 #ifdef HAVE_LZMA
 	#include <lzma.h>
 #endif
+
+#ifdef __APPLE__ // OS X uses <machine/endian.h> not <endian.h>.
+  #include <machine/endian.h>
+	#include <libkern/OSByteOrder.h>
+#else
 #include <endian.h>
+#endif // __APPLE__
+
 #include <stdint.h>

 #include <apti18n.h>
 									/*}}}*/

@@ -1883,11 +1890,15 @@
        if (read(iFd, &size, 4) != 4)
        {
 	  FileFdErrno("read","Unable to read original size of gzipped file");
 	  return 0;
        }
+			 #ifdef __APPLE__
+			 	size = OSSwapLittleToHostInt32(size); // OS X uses different endian conversion functions.
+			 #else
        size = le32toh(size);
+			 #endif // __APPLE__

        if (lseek(iFd, oldPos, SEEK_SET) < 0)
        {
 	  FileFdErrno("lseek","Unable to seek in gzipped file");
 	  return 0;
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/contrib/sha2_internal.cc apt-1.0.9.8.3/apt-pkg/contrib/sha2_internal.cc
--- apt-1.0.9.8.3.original/apt-pkg/contrib/sha2_internal.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/contrib/sha2_internal.cc	2016-06-04 12:49:29.000000000 +0100
@@ -31,11 +31,16 @@
  *
  * $Id: sha2.c,v 1.1 2001/11/08 00:01:51 adg Exp adg $
  */
 #include <config.h>

+#ifdef __APPLE__
+	#include <machine/endian.h>
+#else
 #include <endian.h>
+#endif // __APPLE__
+
 #include <string.h>	/* memcpy()/memset() or bcopy()/bzero() */
 #include <assert.h>	/* assert() */
 #include "sha2_internal.h"

 /*
@@ -1076,6 +1081,5 @@

 	SHA384_Init(&context);
 	SHA384_Update(&context, data, len);
 	return SHA384_End(&context, digest);
 }
-
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/deb/debsrcrecords.cc apt-1.0.9.8.3/apt-pkg/deb/debsrcrecords.cc
--- apt-1.0.9.8.3.original/apt-pkg/deb/debsrcrecords.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/deb/debsrcrecords.cc	2016-06-04 14:02:55.000000000 +0100
@@ -17,10 +17,14 @@
 #include <apt-pkg/strutl.h>
 #include <apt-pkg/aptconfiguration.h>
 #include <apt-pkg/srcrecords.h>
 #include <apt-pkg/tagfile.h>

+#ifdef __APPLE__
+  #include "darwin/strchrnul.h"
+#endif // __APPLE__
+
 #include <ctype.h>
 #include <stdlib.h>
 #include <string.h>
 #include <algorithm>
 #include <string>
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/deb/debversion.cc apt-1.0.9.8.3/apt-pkg/deb/debversion.cc
--- apt-1.0.9.8.3.original/apt-pkg/deb/debversion.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/deb/debversion.cc	2016-06-04 14:00:56.000000000 +0100
@@ -13,10 +13,14 @@
 #include <config.h>

 #include <apt-pkg/debversion.h>
 #include <apt-pkg/pkgcache.h>

+#ifdef __APPLE__
+  #include "darwin/memrchr.h"
+#endif // __APPLE__
+
 #include <string.h>
 #include <string>
 #include <stdlib.h>
 #include <ctype.h>
 									/*}}}*/
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/deb/dpkgpm.cc apt-1.0.9.8.3/apt-pkg/deb/dpkgpm.cc
--- apt-1.0.9.8.3.original/apt-pkg/deb/dpkgpm.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/deb/dpkgpm.cc	2016-06-04 14:01:09.000000000 +0100
@@ -25,11 +25,10 @@
 #include <apt-pkg/pkgcache.h>

 #include <errno.h>
 #include <fcntl.h>
 #include <grp.h>
-#include <pty.h>
 #include <pwd.h>
 #include <signal.h>
 #include <stddef.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -48,20 +47,32 @@
 #include <set>
 #include <string>
 #include <utility>
 #include <vector>

+#ifdef __APPLE__ // OS X stores some functions from <pty.h> under <util.h>.
+  #include <util.h>
+  #include <crt_externs.h> // And while we're sorting out <pty.h> and <util.h>, include <crt_externs.h> for _NSGetEnviron().
+  #include "darwin/memrchr.h"
+#else
+  #include <pty.h>
+#endif // __APPLE__
+
 #include <apti18n.h>
 									/*}}}*/

 using namespace std;

 APT_PURE static unsigned int
 EnvironmentSize()
 {
   unsigned int size = 0;
+  #ifdef __APPLE__
+    char **envp = (char**) _NSGetEnviron(); // We don't have `environ' on OS X so we need to use and cast _NSGetEnviron().
+  #else
   char **envp = environ;
+  #endif // __APPLE__

   while (*envp != NULL)
     size += strlen (*envp++) + 1;

   return size;
@@ -421,11 +432,15 @@
    Configuration::Item const *Opts = _config->Tree(Cnf);
    if (Opts == 0 || Opts->Child == 0)
       return true;
    Opts = Opts->Child;

+   #ifdef __APPLE__
+    sig_t old_sigpipe = signal(SIGPIPE, SIG_IGN);
+   #else
    sighandler_t old_sigpipe = signal(SIGPIPE, SIG_IGN);
+   #endif // __APPLE__

    unsigned int Count = 1;
    for (; Opts != 0; Opts = Opts->Next, Count++)
    {
       if (Opts->Value.empty() == true)
@@ -1507,20 +1522,29 @@

       /* Mask off sig int/quit. We do this because dpkg also does when
          it forks scripts. What happens is that when you hit ctrl-c it sends
 	 it to all processes in the group. Since dpkg ignores the signal
 	 it doesn't die but we do! So we must also ignore it */
+      #ifdef __APPLE__
+        sig_t old_SIGQUIT = signal(SIGQUIT,SIG_IGN);
+        sig_t old_SIGINT = signal(SIGINT,SigINT);
+      #else
       sighandler_t old_SIGQUIT = signal(SIGQUIT,SIG_IGN);
       sighandler_t old_SIGINT = signal(SIGINT,SigINT);
+      #endif // __APPLE__

       // Check here for any SIGINT
       if (pkgPackageManager::SigINTStop && (Op == Item::Remove || Op == Item::Purge || Op == Item::Install))
          break;


       // ignore SIGHUP as well (debian #463030)
+      #ifdef __APPLE__
+        sig_t old_SIGHUP = signal(SIGHUP,SIG_IGN);
+      #else
       sighandler_t old_SIGHUP = signal(SIGHUP,SIG_IGN);
+      #endif // __APPLE__

       // now run dpkg
       d->progress->StartDpkg();
       std::set<int> KeepFDs;
       KeepFDs.insert(fd[1]);
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/install-progress.h apt-1.0.9.8.3/apt-pkg/install-progress.h
--- apt-1.0.9.8.3.original/apt-pkg/install-progress.h	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/install-progress.h	2016-06-04 12:47:57.000000000 +0100
@@ -132,11 +132,15 @@
        int rows;
        int columns;
     } TermSize;
     TermSize GetTerminalSize();

+    #ifdef __APPLE__
+    sig_t old_SIGWINCH;
+    #else
     sighandler_t old_SIGWINCH;
+    #endif // __APPLE__
     int child_pty;

  public:
     PackageManagerFancy();
     ~PackageManagerFancy();
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/makefile apt-1.0.9.8.3/apt-pkg/makefile
--- apt-1.0.9.8.3.original/apt-pkg/makefile	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/makefile	2016-06-04 17:34:10.000000000 +0100
@@ -24,10 +24,17 @@
 endif
 ifeq ($(HAVE_LZMA),yes)
 SLIBS+= -llzma
 endif
 APT_DOMAIN:=libapt-pkg$(LIBAPTPKG_MAJOR)
+#ifeq ($(shell uname), Darwin)
+#SLIBS+= -L$(BASE)/darwin -ldarwinmissing.a
+#endif

 SOURCE = $(wildcard *.cc */*.cc)
 HEADERS = $(addprefix apt-pkg/,$(notdir $(wildcard *.h */*.h)))

+ifeq ($(shell uname), Darwin)
+SOURCE+= $(BASE)/darwin/strchrnul.cc $(BASE)/darwin/memrchr.cc
+endif
+
 include $(LIBRARY_H)
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-pkg/srcrecords.cc apt-1.0.9.8.3/apt-pkg/srcrecords.cc
--- apt-1.0.9.8.3.original/apt-pkg/srcrecords.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-pkg/srcrecords.cc	2016-06-04 12:39:17.000000000 +0100
@@ -29,11 +29,11 @@
 									/*}}}*/

 // SrcRecords::pkgSrcRecords - Constructor				/*{{{*/
 // ---------------------------------------------------------------------
 /* Open all the source index files */
-pkgSrcRecords::pkgSrcRecords(pkgSourceList &List) : d(NULL), Files(0), Current(0)
+pkgSrcRecords::pkgSrcRecords(pkgSourceList &List) : d(NULL), Files(0), Current()
 {
    for (pkgSourceList::const_iterator I = List.begin(); I != List.end(); ++I)
    {
       std::vector<pkgIndexFile *> *Indexes = (*I)->GetIndexFiles();
       for (std::vector<pkgIndexFile *>::const_iterator J = Indexes->begin();
diff -w -r --unified=5 apt-1.0.9.8.3.original/apt-private/private-install.cc apt-1.0.9.8.3/apt-private/private-install.cc
--- apt-1.0.9.8.3.original/apt-private/private-install.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/apt-private/private-install.cc	2016-06-04 16:32:23.000000000 +0100
@@ -22,11 +22,16 @@
 #include <apt-pkg/upgrade.h>

 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
+#ifdef __APPLE__
+  #include <sys/param.h>
+  #include <sys/mount.h>
+#else
 #include <sys/statfs.h>
+#endif // __APPLE__
 #include <sys/statvfs.h>
 #include <algorithm>
 #include <iostream>
 #include <set>
 #include <vector>





diff -w -r --unified=5 apt-1.0.9.8.3.original/buildlib/defaults.mak apt-1.0.9.8.3/buildlib/defaults.mak
--- apt-1.0.9.8.3.original/buildlib/defaults.mak	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/buildlib/defaults.mak	2016-06-04 15:18:48.000000000 +0100
@@ -106,11 +106,11 @@
 # will be published to.
 # This should be declared before including this file
 HEADER_TARGETDIRS+=

 # Options
-CPPFLAGS+= -I$(INCLUDE)
+CPPFLAGS+= -I$(INCLUDE) -I$(SRCDIR) -I$(SRCDIR)/include
 LDFLAGS+= -L$(LIB)

 # Directors to create
 MKDIRS := $(BIN)

diff -w -r --unified=5 apt-1.0.9.8.3.original/buildlib/environment.mak.in apt-1.0.9.8.3/buildlib/environment.mak.in
--- apt-1.0.9.8.3.original/buildlib/environment.mak.in	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/buildlib/environment.mak.in	2016-06-04 15:09:49.000000000 +0100
@@ -8,12 +8,14 @@
 # C++ compiler options
 CC = @CC@
 CPPFLAGS+= @CPPFLAGS@ @DEFS@ -D_REENTRANT -D_FORTIFY_SOURCE=2
 CXX = @CXX@
 CXXFLAGS+= @CXXFLAGS@ -Wall -Wextra
-CXXFLAGS+= -Wcast-align -Wlogical-op -Wredundant-decls -Wmissing-declarations -Wunsafe-loop-optimizations
-CXXFLAGS+= -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=noreturn
+# bad for OS X
+#CXXFLAGS+= -Wcast-align -Wlogical-op -Wredundant-decls -Wmissing-declarations -Wunsafe-loop-optimizations
+#CXXFLAGS+= -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=noreturn
+#####
 # a bit too pedantic to be run by default
 #CXXFLAGS+= -Wpedantic -Wno-long-long -Wno-vla -Wno-variadic-macros
 NUM_PROCS = @NUM_PROCS@

 # Linker stuff
@@ -73,5 +75,7 @@
    LFLAGS_SO=
 else
    # Do not know how to create shared libraries here.
    ONLYSTATICLIBS = yes
 endif
+
+SRCDIR = @abs_srcdir@
diff -w -r --unified=5 apt-1.0.9.8.3.original/cmdline/apt-get.cc apt-1.0.9.8.3/cmdline/apt-get.cc
--- apt-1.0.9.8.3.original/cmdline/apt-get.cc	2016-03-12 13:14:29.000000000 +0000
+++ apt-1.0.9.8.3/cmdline/apt-get.cc	2016-06-04 16:50:09.000000000 +0100
@@ -76,11 +76,16 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/ioctl.h>
 #include <sys/stat.h>
+#ifdef __APPLE__
+  #include <sys/param.h>
+  #include <sys/mount.h>
+#else
 #include <sys/statfs.h>
+#endif // __APPLE__
 #include <sys/statvfs.h>
 #include <sys/wait.h>
 #include <unistd.h>
 #include <algorithm>
 #include <fstream>
